File: apps/cli/src/files.rs
use crate::api_types::DeploymentFile;
use color_eyre::eyre::{Result, eyre};
use std::{fs, path::Path};

pub(crate) fn collect_files(root: &Path) -> Result<Vec<DeploymentFile>> {
    use ignore::WalkBuilder;

    let walker = WalkBuilder::new(root)
        .git_ignore(true)
        .git_global(true)
        .git_exclude(true)
        .ignore(true)
        .hidden(true)
        .follow_links(false)
        .filter_entry(|entry| {
            let dominated_directory =
                |name| entry.file_type().is_some_and(|ft| ft.is_dir()) && entry.file_name() == name;
            !dominated_directory("node_modules")
                && !dominated_directory("target")
                && !dominated_directory("dist")
                && !dominated_directory(".output")
                && !dominated_directory(".next")
                && !dominated_directory(".nuxt")
                && !dominated_directory(".svelte-kit")
                && !dominated_directory("build")
                && !dominated_directory("out")
                && !dominated_directory(".turbo")
                && !dominated_directory(".cache")
                && !dominated_directory("coverage")
                && !dominated_directory(".parcel-cache")
                && !dominated_directory(".vite")
        })
        .build();

    let mut files = Vec::new();
    for result in walker {
        let entry = result.map_err(|err| eyre!("Failed to walk directory: {err}"))?;
        let path = entry.path();
        if path.is_dir() || !is_allowed_extension(path) {
            continue;
        }
        let contents = fs::read_to_string(path)
            .map_err(|err| eyre!("Failed to read {}: {err}", path.display()))?;
        let rel = path_to_relative(path, root)?;
        files.push(DeploymentFile {
            path: rel,
            contents,
        });
    }

    if files.is_empty() {
        return Err(eyre!("No files found under {}", root.display()));
    }
    Ok(files)
}

fn is_allowed_extension(path: &Path) -> bool {
    match path.extension().and_then(|ext| ext.to_str()) {
        Some("ts" | "tsx" | "js" | "jsx" | "mjs" | "cts") => true,
        _ => false,
    }
}

pub(crate) fn path_to_relative(path: &Path, root: &Path) -> Result<String> {
    let rel = path
        .strip_prefix(root)
        .map_err(|_| eyre!("Entry file is not inside {}", root.display()))?;
    let rel = rel.to_string_lossy().replace('\\', "/");
    if rel.is_empty() {
        return Err(eyre!("Entry path is empty"));
    }
    Ok(rel)
}

File: apps/cli/src/lib/files.ts
import { readFile } from 'node:fs/promises'
import path from 'node:path'

import ignore from 'ignore'
import { glob } from 'tinyglobby'

export type DeploymentFile = {
  path: string
  contents: string
}

const ALLOWED_EXTENSIONS = new Set(['.ts', '.tsx', '.js', '.jsx', '.mjs', '.cts'])
const EXTRA_FILE_NAMES = new Set([
  'package.json',
  'pnpm-lock.yaml',
  'package-lock.json',
  'yarn.lock',
  'bun.lockb',
  'tsconfig.json'
])
const SKIP_DIRS = new Set([
  'node_modules',
  'target',
  'dist',
  '.output',
  '.next',
  '.nuxt',
  '.svelte-kit',
  'build',
  'out',
  '.turbo',
  '.cache',
  'coverage',
  '.parcel-cache',
  '.vite',
  '.git'
])

export async function collectFiles(root: string): Promise<DeploymentFile[]> {
  const rootAbs = path.resolve(root)
  const ignorePatterns = [...SKIP_DIRS].map((dir) => `**/${dir}/**`)
  const includePatterns = [
    ...[...ALLOWED_EXTENSIONS].map((ext) => `**/*${ext}`),
    ...[...EXTRA_FILE_NAMES].map((fileName) => `**/${fileName}`)
  ]
  const relPaths = await glob(includePatterns, {
    cwd: rootAbs,
    dot: true,
    onlyFiles: true,
    ignore: ignorePatterns,
    followSymbolicLinks: false
  })

  const ignoreMatcher = await buildIgnoreMatcher(rootAbs, ignorePatterns)

  const files: DeploymentFile[] = []
  for (const rel of relPaths) {
    if (ignoreMatcher.ignores(rel)) {
      continue
    }

    const abs = path.join(rootAbs, rel)
    const contents = await readFile(abs, 'utf8')
    files.push({ path: rel.replace(/\\/g, '/'), contents })
  }

  if (files.length === 0) {
    throw new Error(`No files found under ${rootAbs}`)
  }

  return files
}

async function buildIgnoreMatcher(rootAbs: string, ignorePatterns: string[]) {
  const matcher = ignore()
  const ignoreFiles = await glob(['.gitignore', '.ignore', '**/.gitignore', '**/.ignore'], {
    cwd: rootAbs,
    dot: true,
    onlyFiles: true,
    ignore: ignorePatterns,
    followSymbolicLinks: false
  })

  ignoreFiles.sort((a, b) => depth(a) - depth(b) || a.localeCompare(b))

  for (const rel of ignoreFiles) {
    const content = await readFile(path.join(rootAbs, rel), 'utf8')
    const dir = path.posix.dirname(rel)
    const prefix = dir === '.' ? '' : `${dir}/`

    const patterns = content
      .split('\n')
      .map((line) => line.trim())
      .filter((line) => line && !line.startsWith('#'))
      .map((pattern) => toRootPattern(pattern, prefix))
      .filter((pattern): pattern is string => Boolean(pattern))

    matcher.add(patterns)
  }

  return matcher
}

function toRootPattern(rawPattern: string, prefix: string): string | null {
  const negated = rawPattern.startsWith('!')
  let pattern = negated ? rawPattern.slice(1) : rawPattern
  if (!pattern) {
    return null
  }

  const rooted = pattern.startsWith('/')
  if (rooted) {
    pattern = pattern.slice(1)
  }
  if (!pattern) {
    return null
  }

  const withTree = pattern.endsWith('/') ? `${pattern}**` : pattern
  const fullPattern = rooted || withTree.includes('/')
    ? `${prefix}${withTree}`
    : `${prefix}**/${withTree}`

  return negated ? `!${fullPattern}` : fullPattern
}

function depth(relPath: string): number {
  return relPath.split('/').length
}

export function toRelative(filePath: string, root: string): string {
  const rel = path.relative(root, filePath).replace(/\\/g, '/')

  if (!rel || rel.startsWith('..')) {
    throw new Error(`Entry file is not inside ${root}`)
  }

  return rel
}
